samples/bluetooth/peripheral_ancs_client/src/main.c

```diff
@@ -105,55 +114,119 @@ static const char *lit_attrid[BT_ANCS_NOTIF_ATTR_COUNT] = {
-		err = bt_ancs_subscribe_notification_source(&ancs_c,
-				bt_ancs_notification_source_handler);
-		if (err) {
-			printk("Failed to enable Notification Source notification (err %d)\n",
-			       err);
-		}
+	err = bt_ancs_subscribe_notification_source(ancs_c,
+			bt_ancs_notification_source_handler);
+	if (err) {
+		printk("Failed to enable Notification Source notification (err %d)\n",
+		       err);
+	}
 
-		err = bt_ancs_subscribe_data_source(&ancs_c,
-				bt_ancs_data_source_handler);
-		if (err) {
-			printk("Failed to enable Data Source notification (err %d)\n",
-			       err);
-		}
+	err = bt_ancs_subscribe_data_source(ancs_c,
+			bt_ancs_data_source_handler);
+	if (err) {
+		printk("Failed to enable Data Source notification (err %d)\n",
+		       err);
 	}
 }
 
-static void discover_completed_cb(struct bt_gatt_dm *dm, void *ctx)
+static void discover_ancs_completed_cb(struct bt_gatt_dm *dm, void *ctx)
 {
 	int err;
+	struct bt_ancs_client *ancs_c = (struct bt_ancs_client *)ctx;
+	struct bt_conn *conn = bt_gatt_dm_conn_get(dm);
 
-	printk("The discovery procedure succeeded\n");
+	printk("The discovery procedure for ANCS succeeded\n");
 
 	bt_gatt_dm_data_print(dm);
 
-	err = bt_ancs_handles_assign(dm, &ancs_c);
+	err = bt_ancs_handles_assign(dm, ancs_c);
 	if (err) {
 		printk("Could not init ANCS client object, error: %d\n", err);
 	} else {
-		has_ancs = true;
+		atomic_set_bit(&discovery_flags, DISCOVERY_ANCS_SUCCEEDED);
+		enable_ancs_notifications(ancs_c);
+	}
 
-		if (bt_conn_get_security(ancs_c.conn) < BT_SECURITY_L2) {
-			err = bt_conn_set_security(ancs_c.conn, BT_SECURITY_L2);
-			if (err) {
-				printk("Failed to set security (err %d)\n",
-				       err);
-			}
-		} else {
-			enable_notifications();
-		}
+	err = bt_gatt_dm_data_release(dm);
+	if (err) {
+		printk("Could not release the discovery data, error "
+		       "code: %d\n",
+		       err);
+	}
+
+	atomic_clear_bit(&discovery_flags, DISCOVERY_ANCS_ONGOING);
+	discover_ancs_2(conn);
+}
+
+static void discover_ancs_not_found_cb(struct bt_conn *conn, void *ctx)
+{
+	printk("ANCS could not be found during the discovery\n");
+
+	atomic_clear_bit(&discovery_flags, DISCOVERY_ANCS_ONGOING);
+	discover_ancs_2(conn);
+}
+
+static void discover_ancs_error_found_cb(struct bt_conn *conn, int err, void *ctx)
+{
+	printk("The discovery procedure for ANCS failed, err %d\n", err);
+
+	atomic_clear_bit(&discovery_flags, DISCOVERY_ANCS_ONGOING);
+	discover_ancs_2(conn);
+}
+
+static const struct bt_gatt_dm_cb discover_ancs_cb = {
+	.completed = discover_ancs_completed_cb,
+	.service_not_found = discover_ancs_not_found_cb,
+	.error_found = discover_ancs_error_found_cb,
+};
+
+static void indicate_sc_cb(struct bt_gatts_client *gatts_c,
+			   const struct bt_gatts_handle_range *handle_range,
+			   int err)
+{
+	if (!err) {
+		atomic_set_bit(&discovery_flags, SERVICE_CHANGED_INDICATED);
+		discover_ancs_2(gatts_c->conn);
```

Comment:
Discovery Manager does not support handle range yet. It will involve more code change.

When ANCS was not found and there is a Service Changed indication, we try to discover ANCS one more time. It is not as optimal as handle range but it should also work. nRF5 SDK example also does not specify handle range.